### ISSUE TITLE
Suggestions for improvements to core drawing functionality

### ISSUE THREAD

**Issue by liuguangxi** [typst/typst#4123]
### Description

With the long-term use of the Typst drawing function, some functional or performance shortcomings gradually appeared. The famous drawing libraries TikZ and PSTricks in the LaTeX world, are good reference. Suggested improvements to the core drawing functionality are detailed below. I expect some of the functionality to be difficult to implement or cause breaking changes in interfaces, but they are worth considering as a direction for future improvements.

### `vec2`

`vec2` is a pair of real numbers, which can be seen as a point or a vector on a plane, or as a complex number. The `vec2` type and related operations are frequently used in various graphics processing, so the built-in implementation of them can help improve performance. Typical usage examples might be as follows:

```typ
#let v = vec2(1.5, 3.0)         // create a `vec2` object

#(vec2(1, 2) + vec2(3, 4))      // addition
#(vec2(1, 2) * 4.5)             // scale
#vec2(3.0, 4.0).length()        // length of vector
#vec2(3.0, 4.0).angle()         // angle of vector
#vec2(3.0, 4.0).dot(v)          // dot of two vectors
#v.min()                        // minimum of two elements in vector
#v.max()                        // maximum of two elements in vector
```

### `transform`

`transform` implement affine transform in 2D geometry. Internally, it is denoted by `t = (t.x, t.y, t.xx, t.xy, t.yx, t.yy)`. A point `(x, y)` is transformed by the transform t to `(x', y')`, where
```
x' = t.x + t.xx * x + t.xy * y
y' = t.y + t.yx * x + t.yy * y
```
Transforms can be applied to pairs, paths and graphics by multiplication on the left. And they can also be composed with one another and inverted.

Constructor for `transform` object may be:

```typ
#let transform.identity() -> transform                      // identity transform
#let transform.translate(dx, dy) -> transform               // translate transform
#let transform.scale(sx, sy) -> transform                   // scale transform
#let transform.rotate(angle, pt) -> transform               // rotate transform by angle about point pt
#let transform.matrix(x, y, xx, xy, yx, yy) -> transform    // general transform
```

The functions for `transform` object might be as follows (suppose `t1` and `t2` are `transform` objects):

```typ
#t1.mat()                   // internal matrix elements of transform
#t1.lmul(t2)                // transform multiplication
#t1.inv()                   // inverse of transform
#vec2(1.5, 2.0).lmul(t1)    // transform vector using t1
```

### `path`

Currently Typst does not have an explicit path object for user, instead the path objects is tightly coupled to the drawing function (implemented as `path` function). This approach is not flexible enough and introduce some troubles for future extention. Based on the idea that path definition is separated from drawing, as in mainstream drawing library. A dedicated `path` type could be defined.

A path is specified as a list of points interconnected with straight line segments or cubic spline segments, and it can be closed or open. Constructor function for `path` object may be:

```typ
#let path(closed: [true | false], vertices) -> path
```

Where `vertices` is array of vertices of the path. Each vertex can be defined in 2 ways:
- A regular point (`vec2` type). Create a straight line segment with previous vertex.
- An array of three points (array of `vec2` type), the first two being the control points, and the last one being the vertex. Create a cubic spline segment with previous vertex.

Several paths can be group into a path array, i.e. array of `path`, as a compound path. For example, a rectangle with a hole in it.
```typ
#let p-square = path(...)           // path of square
#let p-circle = path(...)           // path of circle
#let ps = (p-square, p-circle)      // compound path
```

The functions for `path` object might be as follows:

```typ
#let p1 = path(
    closed: true,
    ( vec2(0, 0), (vec2(4, 2), vec2(6, 2), vec2(10, 0)) )
)    // create a closed path

#p1.closed()                    // whether path is closed
#p1.vertices()                  // array of vertices of the path
#p1.size()                      // the number of nodes in the path
#p1.reverse()                   // return reversed path
#p1.min()                       // pair (left,bottom) for the bounding box of path
#p1.max()                       // pair (right,top) for the bounding box of path
#p1.lmul(t1)                    // return transformed path using t1
#p1.sub-path(a, b)              // return the subpath of p1 running from node a to node b
#path.merge-path(p1, p2, p3)    // return merged path p1, p2, p3
```

Path transformation operations are computationally intensive and frequently present in graphics processing, so the built-in implementation of them can help improve performance.

More path-related functions can be implemented by the extension library.

## Refine drawing functions

The following drawing functions need to be provided. They are all path-based.

```typ
#let draw(draw-pen: [`color` | ...], [`path` | array of `path`])

#let fill(fill-pen: [`color` | ...], fill-rule: ["nonzero" | "evenodd"], [`path` | array of `path`])

#let filldraw(fill-pen: [`color` | ...], draw-pen: [`color` | ...], fill-rule: ["nonzero" | "evenodd"], [`path` | array of `path`])

#let clip(clip-rule: ["nonzero" | "evenodd"], [`path` | array of `path`])
```

The first three functions stroke and/or fill path(compound path) using predefined color, gradient, etc. Note that both **nonzero** and **evenodd** fill rules should be supported (except in function `draw`).

The last function `clip` implements cliping the current drawing contents to the region bounded by the path(compound path), where both **nonzero** and **evenodd** clip rules are supported. This function may be complicated to implement, and the scope of the clipping operation needs to be more sensibly defined.

These drawing functions are already available in the library TikZ, PSTricks and in SVG specification, so Typst should consider refine these in the future.

### Use Case

The following graphics would be relatively easy to draw if Typst supports the previously mentioned improvements.

A path of a circle with 4 holes in it fills the gradient, and lay on top of the texts.

![tst](https://github.com/typst/typst/assets/2703168/65114d2d-36bb-4aad-9018-9d264f014ff5)

**Comment by bluebear94** [comment]
Note that circles and ellipses can’t be reproduced exactly with Bézier curves, but Typst [already approximates them anyway](https://github.com/typst/typst/blob/main/crates/typst/src/visualize/shape.rs#L609).

**Comment by liuguangxi** [comment]
Indeed, circles, ellipses, arcs, and elliptical arcs can all be approximated using no more than 4 cubic Bézier curves.

**Comment by joneuhauser** [comment]
Not having a clip rule / compound path is a serious limitation, which currently prevents me from implementing [the plt donuts example](https://matplotlib.org/stable/gallery/shapes_and_collections/donut.html#sphx-glr-gallery-shapes-and-collections-donut-py) in https://github.com/daskol/typst-mpl-backend

**Comment by knuesel** [comment]
@NeilGirdhar the situation with Typst is very different from LaTeX. It's one of the biggest differences between the two systems: LaTeX is horrible to program, while Typst includes a full-fledged scripting language that is a joy to use. At least I find it more pleasant than Python for expressing drawings. (And if you really want to make a library in another programming language, that's possible with the [plugin](https://typst.app/docs/reference/foundations/plugin/) system.)

As for a library like TikZ, there is already the [CeTZ](https://typst.app/universe/package/cetz/) package. It's written in Typst script and is already very usable, though not yet as mature as TikZ.

I think this GitHub issue is about adding more primitives in the Typst language to make packages like CeTZ easier to write and more efficient.

**Comment by NeilGirdhar** [comment]
> It's one of the biggest differences between the two systems: LaTeX is horrible to program,

Agreed.  No shade on the LaTeX/Tikz programmers who have put a huge amount of effort into a difficult situation.

>  Typst includes a full-fledged scripting language that is a joy to use. At least I find it more pleasant than Python for expressing drawings.

Okay, I'll take your word for it.

> As for a library like TikZ, there is already the [CeTZ](https://typst.app/universe/package/cetz/) package. It's written in Typst script and is already very usable, though not yet as mature as TikZ.

Are the drawings truly incrementally compiled?  I.e., if the drawing code doesn't change, is the drawing recompiled?

> I think this GitHub issue is about adding more primitives in the Typst language to make packages like CeTZ easier to write and more efficient.

I see.   My fundamental point is that, IMO LaTeX and Tikz made the mistake of trying to be _servers_ that have programs written for them.  But neither is a very good server.   They should have been _clients_, i.e., libraries that client code calls into .   This would allow clients to have limitless flexibility in terms of language, data structures, etc.  Clients might say something like: here's a paragraph, typeset it with this context, and give me the result.  And then the client could make another call and say, create a box with arrows, and place the text that we typeset into it, which produces another result.  And then the client could say, okay now put that in the center of the page.  Essentially, I've always felt limited by LaTeX.

So in answer to your answer, I'm not looking for more primitives.  My instict is that CeTZ should ideally be written in Rust and interact with Typest using a small, optimized interface.  But maybe you're saying that the interface is too small, and new functions are needed, in which case, I guess we agree.

**Comment by knuesel** [comment]
> Are the drawings truly incrementally compiled? I.e., if the drawing code doesn't change, is the drawing recompiled?

Yes! (well it gets recompiled if the context changes, e.g. if you change the font or other style that is used in the drawing). It's very handy when you're working on a figure and want to adjust a coordinate or an angle, you can get the real-time preview of the result in another window on the side.

Regarding your point about clients: Typst provides a Rust crate you can use to call the incremental compiler as a library, so you can embed Typst in your own Rust program. Maybe that would help with your preferred workflow... 

> My instict is that CeTZ should ideally be written in Rust and interact with Typest using a small, optimized interface.

I guess that would be nice to get better performance, but it would come at the cost of higher barrier to entry for contributors. Currently if CeTZ doesn't implement the kind of plot you need, you can write a few lines of Typst in your document to do it yourself (compare that to forking the hypothetical Rust version ofCeTZ, installing the Rust toolchain, etc.)

One of the great things with Typst is that you can do pretty much everything as a simple document. It's self-contained: just call `typst compile file.typ` and the compiler will download any necessary package/plugin, it just works on all platforms without extra steps.  And you can do it all in the Typst language. I like this simplicity compared to other solutions that mix Markdown, Python, YAML, LaTeX, etc.

> But maybe you're saying that the interface is too small, and new functions are needed, in which case, I guess we agree.

Yes, I guess we do :)

**Comment by NeilGirdhar** [comment]
>  Typst provides a Rust crate you can use to call the incremental compiler as a library, so you can embed Typst in your own Rust program. Maybe that would help with your preferred workflow...

It just lets me know that the Typest team has chosen a good design, which motivates me to switch to Typest 😄 

> Yes! (well it gets recompiled if the context changes, e

Right, but what  I'm asking: if the content of the image doesn't change, but the text around it changes, is the image recompiled?

> One of the great things with Typst is that you can do pretty much everything as a simple document. It's self-contained: just call typst compile file.typ 

I understand, but this desire for self-containment is exactly how Latex got into trouble.  They spawned all kinds of hacks, like inputting other files (so that your document is essentially a monolith) and then they couldn't do incremental compilation, so that you have rebuild everything unless you embark on other hacks.

Also trying to do everything in a document is handy for small documents, but most people writing Latex/Typest have big documents, so I think it's important to focus on how easy it is to do complex things rather than how simple the simplest programs are.

Anyway, I'm definitely keeping an eye on Typest.  It would be cool if there were an issue tag for wherever the typesetting is inferior to latex so that users can keep track of them as they get fixed 😄 .  I can find workarounds for missing features, but not bad typesetting.
